apollo-router/src/services/router/service.rs

```diff
@@ -88,38 +86,54 @@ static ACCEL_BUFFERING_HEADER_NAME: HeaderName = HeaderName::from_static("x-acce
 static ACCEL_BUFFERING_HEADER_VALUE: HeaderValue = HeaderValue::from_static("no");
 
 /// Containing [`Service`] in the request lifecyle.
-#[derive(Clone)]
 pub(crate) struct RouterService {
-    apq_layer: Arc<APQLayer>,
-    persisted_query_layer: Arc<PersistedQueryLayer>,
-    query_analysis_layer: Arc<QueryAnalysisLayer>,
-    // Cannot be under Arc. Batching state must be preserved for each RouterService
-    // instance
-    batching: Batching,
-    supergraph_service: supergraph::BoxCloneService,
-    apollo_telemetry_config: ApolloTelemetryConfig,
+    // A service stack for the actual implementation of the router service.
+    service: router::BoxService,
 }
 
 impl RouterService {
     fn new(
-        sgb: supergraph::BoxService,
+        supergraph_service: supergraph::BoxService,
         apq_layer: APQLayer,
         persisted_query_layer: Arc<PersistedQueryLayer>,
         query_analysis_layer: QueryAnalysisLayer,
         batching: Batching,
         apollo_telemetry_config: ApolloTelemetryConfig,
     ) -> Self {
-        let supergraph_service: supergraph::BoxCloneService =
-            ServiceBuilder::new().buffered().service(sgb).boxed_clone();
+        // Some of the layers in the stack are wrapping previous implementations that are called
+        // layers, but are not tower layers at all.
+        let apq_layer = Arc::new(apq_layer);
+        let query_analysis_layer = Arc::new(query_analysis_layer);
+
+        let service = ServiceBuilder::new()
+            .layer(DisplayRouterRequestLayer)
```

Comment:
This layer is re-ordered: previously, it happened in the middle of other batching/router service code. It is not possible to split up batching from the router service without changing the ordering of this feature.

It must be before the batching layer, because it has to log the entire batch request as one, not as multiple single-query requests.

It reads the body, but only if it actually needs to.